packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/documents_batch/transformer/v0/mod.rs

```diff
@@ -0,0 +1,466 @@
+    ) -> Result<ConsensusValidationResult<Vec<DocumentTransitionAction>>, Error> {
+        let drive = platform.drive;
+        // Data Contract must exist
+        let Some(data_contract_fetch_info) = drive
+        .get_contract_with_fetch_info_and_fee(data_contract_id.0 .0, None, false, transaction, platform_version)?
+        .1
+        else {
+            return Ok(ConsensusValidationResult::new_with_error(BasicError::DataContractNotPresentError(DataContractNotPresentError::new(*data_contract_id)).into()));
+        };
+
+        let validation_result = document_transitions
+        .iter()
+        .map(|(document_type_name, document_transitions)| {
+            Self::transform_document_transitions_within_document_type_v0(
+                platform,
+                validate,
+                data_contract_fetch_info.clone(),
+                document_type_name,
+                owner_id,
+                document_transitions,
+                execution_context,
+                transaction,
+                platform_version,
+            )
+        })
+        .collect::<Result<Vec<ConsensusValidationResult<Vec<DocumentTransitionAction>>>, Error>>(
+        )?;
+        Ok(ConsensusValidationResult::flatten(validation_result))
+    }
+
+    fn transform_document_transitions_within_document_type_v0(
+        platform: &PlatformStateRef,
+        validate: bool,
+        data_contract_fetch_info: Arc<DataContractFetchInfo>,
+        document_type_name: &str,
+        owner_id: Identifier,
+        document_transitions: &[&DocumentTransition],
+        _execution_context: &mut StateTransitionExecutionContext,
+        transaction: TransactionArg,
+        platform_version: &PlatformVersion,
+    ) -> Result<ConsensusValidationResult<Vec<DocumentTransitionAction>>, Error> {
+        // // We use temporary execution context without dry run,
+        // // because despite the dryRun, we need to get the
+        // // data contract to proceed with following logic
+        // let tmp_execution_context = StateTransitionExecutionContext::default_for_platform_version(platform_version)?;
+        //
+        // execution_context.add_operations(tmp_execution_context.operations_slice());
+
+        let dry_run = false; //maybe reenable
+
+        let data_contract = &data_contract_fetch_info.contract;
+
+        let document_type = data_contract.document_type_for_name(document_type_name)?;
+
+        let replace_transitions = document_transitions
+            .iter()
+            .filter(|transition| matches!(transition, DocumentTransition::Replace(_)))
+            .copied()
+            .collect::<Vec<_>>();
+
+        // we fetch all documents needed for the transitions
+        // for create they should not exist
+        // for replace/patch they should
+        // for delete they should
+        // Validation will come after, but doing one request can be faster.
+        let fetched_documents_validation_result =
```

Comment:
We fetch twice?

Reply:
Comment above is wrong. We need to explain here that we fetch only replace and do state validation for replace because we call transform before  state validation